onnxruntime/python/onnxruntime_pybind_state.cc

```diff
@@ -905,105 +1012,287 @@ void addObjectMethods(py::module& m, Environment& env) {
+          // TODO: Add check to ensure that string arrays are not passed - we currently don't support string tensors in CUDA
+          CreateGenericMLValue(nullptr, GetCudaAllocator(device.Id()), "", array_on_cpu, ml_value.get(), true, false, CpuToCudaMemCpy);
+
+#else
+        throw std::runtime_error(
+            "Can't allocate memory on the CUDA device using this package of OnnxRuntime. "
+            "Please use the CUDA package of OnnxRuntime to use this feature.");
+#endif
+        } else {
+          throw std::runtime_error("Unsupported device: Cannot place the OrtValue on this device");
+        }
+
+        return ml_value;
+      })
+
+      // Factory method to create an OrtValue (Tensor) from the given shape and element type with memory on the specified device
+      // The memory is left uninitialized
+      .def_static("tensor_from_shape_and_type", [](std::vector<int64_t>& shape, py::object& element_type, OrtDevice& device) {
+        PyArray_Descr* dtype;
+        if (!PyArray_DescrConverter(element_type.ptr(), &dtype)) {
+          throw std::runtime_error("Not a valid numpy type");
+        }
+
+        int type_num = dtype->type_num;
+        Py_DECREF(dtype);
+
+        if (!IsNumericNumpyType(type_num)) {
+          throw std::runtime_error("Creation of OrtValues is currently only supported from non-string numpy arrays");
+        }
+
+        auto ml_value = onnxruntime::make_unique<OrtValue>();
+
+        std::unique_ptr<Tensor> tensor;
+        // The tensor's memory is allocated on the CPU
+        if (GetDeviceName(device) == CPU) {
+          tensor = onnxruntime::make_unique<Tensor>(NumpyTypeToOnnxRuntimeType(type_num), shape, GetAllocator());
+        } else if (GetDeviceName(device) == CUDA) {
+      // The tensor's memory is allocated on CUDA
+#ifdef USE_CUDA
+          if (!IsCudaDeviceIdValid(logging::LoggingManager::DefaultLogger(), device.Id())) {
+            throw std::runtime_error("The provided device id doesn't match any available GPUs on the machine.");
+          }
+
+          tensor = onnxruntime::make_unique<Tensor>(NumpyTypeToOnnxRuntimeType(type_num), shape, GetCudaAllocator(device.Id()));
+#else
+        throw std::runtime_error(
+            "Can't allocate memory on the CUDA device using this package of OnnxRuntime. "
+            "Please use the CUDA package of OnnxRuntime to use this feature.");
+#endif
+        } else {
+          throw std::runtime_error("Unsupported device: Cannot place the OrtValue on this device");
+        }
+
+        auto ml_tensor = DataTypeImpl::GetType<Tensor>();
+        ml_value->Init(tensor.release(),
+                       ml_tensor,
+                       ml_tensor->GetDeleteFunc());
+
+        return ml_value;
+      })
+      .def("data_ptr", [](OrtValue* ml_value) -> int64_t {
+        // TODO: Assumes that the OrtValue is a Tensor, make this generic to handle non-Tensors
```

Comment:
Why do we want to give out the internal data pointer? In the test python the only usage is in IOBinding, so would it would be better for that API to take an OrtValue instead?

Reply:
Gives the user the option of tapping into the raw data buffer backing the OrtValue. FWIW: Torch Tensor does support this too: https://pytorch.org/docs/stable/tensors.html#torch.Tensor.data_ptr. We (have already) exposed an API to bind raw memory buffers and given that we expose that, it may seem a little weird that we don't have any way to get hold of a pointer to a memory buffer and that the user would have to resort to using a custom allocator for this.